packages/entity-server/src/routes/hierarchy/relations/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,136 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { Route } from '@tupaia/server-boilerplate';
+import { reduceToDictionary, reduceToArrayDictionary } from '@tupaia/utils';
+import { EntityType } from '../../../models';
+import { formatEntitiesForResponse } from '../format';
+import { RelationsRequest } from './types';
+
+type Pair = {
+  descendant: string;
+  ancestor: string;
+};
+
+export class EntityRelationsRoute extends Route<RelationsRequest> {
+  async buildAncestorCodesAndPairs(descendants: EntityType[]): Promise<[string[], Pair[]]> {
+    const { hierarchyId } = this.req.ctx;
+    const { type: ancestorType } = this.req.ctx.ancestor;
+    const descendantAncestorMapping = await this.req.models.entity.fetchAncestorDetailsByDescendantCode(
+      descendants.map(descendant => descendant.code),
+      hierarchyId,
+      ancestorType,
+    );
+    const ancestorCodes = Object.values(descendantAncestorMapping).map(ancestor => ancestor.code);
+    const ancestorDescendantPairs = Object.entries(descendantAncestorMapping).map(
+      ([descendant, { code: ancestor }]) => ({
+        descendant,
+        ancestor,
+      }),
+    );
+
+    return [ancestorCodes, ancestorDescendantPairs];
+  }
+
+  async getAncestors(ancestorsWithDescendantsCodes: string[]) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+    const { type: ancestorType, filter } = this.req.ctx.ancestor;
+    return entity.type === ancestorType
+      ? [entity]
+      : entity.getDescendants(hierarchyId, {
+          code: ancestorsWithDescendantsCodes,
+          ...filter,
+          country_code: allowedCountries,
```

Comment:
Question: if the client provides a `country_code` filter, we should use the intersection of `filter.country_code` and `allowedCountries` as the `country_code` condition. I am not sure if this logic is already used to evaluate `allowedCountries`?

Reply:
That's a good point, we probably ought to 👍

Reply:
It's a bit of a detail and not a big blocker, so happy for it to be addressed in the end, after the refactoring & merging dust has settled